pkg/metrics/gatherer/gatherer.go

```diff
@@ -0,0 +1,141 @@
+		g.pushClient = pushclient.PushClient{URL: prometheusURL, Client: &http.Client{}, JobName: "openshift_installer_metrics"}
+	}
+	g.metricRegistry = make(map[string]*builder.MetricBuilder)
+
+}
+
+// AddLabelValue adds the label key and value to the storage of the Gatherer object.
+func (g *Gatherer) AddLabelValue(metricName string, labelName string, labelValue string) {
+	if enableMetrics {
+		if _, found := g.metricRegistry[metricName]; !found {
+			opts := optsRegistry[metricName]
+			metricBuilder, err := builder.NewMetricBuilder(*opts, 0, nil)
+			if err == nil {
+				g.metricRegistry[metricName] = metricBuilder
+			}
+		}
+		g.metricRegistry[metricName].AddLabelValue(labelName, labelValue)
+	}
+}
+
+// SetValue sets the value of the metric that is stored in the Gatherer object.
+func (g *Gatherer) SetValue(metricName string, value float64) {
+	if enableMetrics {
+		if _, found := g.metricRegistry[metricName]; !found {
+			opts := optsRegistry[metricName]
+			metricBuilder, err := builder.NewMetricBuilder(*opts, 0, nil)
+			if err == nil {
+				g.metricRegistry[metricName] = metricBuilder
+			}
+		}
+		g.metricRegistry[metricName].SetValue(value)
+	}
+}
+
+// Push is a driver function that sends the metrics created with all the information and pushes it
+// to Prometheus through the Push Client.
+func (g *Gatherer) Push(metricName string) {
+	if enableMetrics {
+		if _, found := g.metricRegistry[metricName]; found {
+			collector, err := g.metricRegistry[metricName].PromCollector()
+			if err == nil {
+				g.pushClient.Push(collector)
+			}
+
+		}
+	}
+}
+
+// PushAll function takes all the metrics whose label values are set and pushes them to Prometheus.
+func (g *Gatherer) PushAll() {
+	if enableMetrics {
+		var collectors []prometheus.Collector
+		for _, value := range g.metricRegistry {
+			collector, err := value.PromCollector()
+			if err == nil {
+				collectors = append(collectors, collector)
+			}
+
+		}
+		g.pushClient.Push(collectors...)
+	}
+}
+
+var (
+	// gatherer is the global Gatherer object that will be used to store all the information if this
+	// package's global functions are called for collection of metrics.
+	gatherer Gatherer
+
+	// CreateCommandMetricName is the map of the targets to the respective metric names. Installer can
+	// look this up for the correct metric name given a target.
+	CreateCommandMetricName = map[string]string{
+		"cluster":          createMetricName,
+		"ignition-configs": ignitionMetricName,
+		"install-config":   installConfigMetricName,
+		"manifests":        manifestsMetricName,
+	}
+
+	// CurrentInvocationContext is used to store the current Invocation metric name used for building context.
+	// Invocation data information is gathered from different parts of the installer and it gets hard for the
+	// installer to keep track of the current command and target that is being run. Hence, the current invocation
+	// metric name is stored here in this variable.
+	CurrentInvocationContext string
```

Comment:
Should individual places need to know the context?? should we configure the gatherer with one context in main where we run the command ??

Reply:
Different functions try to update the invocation metrics and it's hard to find out which one to target so I put a common Context with the name to target which they can use. We could put it up in the main function but I thought it would be best left in the metrics package. Currently there is a need of only one context.